### PR TITLE
Transform constraints to z3 constraints which is the final step

### PR DESCRIPTION
**Pull request by migeed-z**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80744
* #80147
* __->__ #80110

Differential Revision: [D37387508](https://our.internmc.facebook.com/intern/diff/D37387508)

We transform our constraints in a way that can be understood by z3. For example, for a dimension type, instead of Dyn, we encode it as a pair of integers. Some of the other operations are a 1:1 transformation.

contains tests for alexnet and resnet as well as smaller examples

### CODE REVIEW

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37387508).

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37387508).

**Review by anijain2305 — APPROVED** [review]

File: test/test_fx.py [hunk 1]
```diff
@@ -45,6 +45,7 @@
 
 if sys.version_info >= (3, 7):
     from fx.test_gradual_type import AnnotationsTest  # noqa: F401
+
```
Comment (jamesr66a) [inline test/test_fx.py:48]: Delete?

**Review by jamesr66a — APPROVED** [review]
This is a lot of code, but I like the test coverage so approving. I trust that this will be validated more in the future with the e2e cases

**Pull request closed without merging**